src/optimizer/Optimizer.cpp

```diff
@@ -0,0 +1,106 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/Optimizer.h"
+
+#include "context/QueryContext.h"
+#include "optimizer/OptGroup.h"
+#include "optimizer/OptRule.h"
+#include "planner/ExecutionPlan.h"
+#include "planner/Logic.h"
+#include "planner/PlanNode.h"
+
+using nebula::graph::BiInputNode;
+using nebula::graph::Loop;
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+using nebula::graph::Select;
+using nebula::graph::SingleDependencyNode;
+
+namespace nebula {
+namespace opt {
+
+Optimizer::Optimizer(QueryContext *qctx, const RuleSet *ruleSet) : qctx_(qctx), ruleSet_(ruleSet) {
+    DCHECK(qctx != nullptr);
+    DCHECK(ruleSet != nullptr);
+}
+
+StatusOr<const PlanNode *> Optimizer::findBestPlan(PlanNode *root) {
+    DCHECK(root != nullptr);
+    rootNode_ = root;
+    NG_RETURN_IF_ERROR(prepare());
+    NG_RETURN_IF_ERROR(doExploration());
+    return rootGroup_->getPlan();
+}
+
+Status Optimizer::prepare() {
+    visitedNodes_.clear();
+    rootGroup_ = convertToGroup(rootNode_);
+    return Status::OK();
+}
+
+Status Optimizer::doExploration() {
+    // TODO(yee): Apply all rules recursively, not only once round
+    for (auto rule : ruleSet_->rules()) {
+        if (!rootGroup_->isExplored(rule)) {
+            NG_RETURN_IF_ERROR(rootGroup_->explore(rule));
+        }
+    }
+    return Status::OK();
+}
+
```

Comment:
How to set the default rule when `enable_optimizer=false`?

Reply:
I'll control rule set selection in query instance. pls wait a minute.

Reply:
Now you can add opt rule to `RuleSet::defaultRules()` like [PushFilterDownGetNbrsRule](https://github.com/vesoft-inc/nebula-graph/pull/247/files#diff-1628cc036fa9a70c831e51ce2cb0884fR32)